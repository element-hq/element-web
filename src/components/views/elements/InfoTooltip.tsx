/*
Copyright 2019 Michael Telatynski <7t3chguy@gmail.com>
Copyright 2019 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React, { ReactNode } from "react";
import classNames from "classnames";
import { Tooltip } from "@vector-im/compound-web";

import { _t } from "../../../languageHandler";

export enum InfoTooltipKind {
    Info = "info",
    Warning = "warning",
}

interface TooltipProps {
    tooltip?: string;
    className?: string;
    kind?: InfoTooltipKind;
    children?: ReactNode;
    tabIndex?: number;
}

export default class InfoTooltip extends React.PureComponent<TooltipProps> {
    public render(): React.ReactNode {
        const { tooltip, children, className, kind } = this.props;
        const title = _t("info_tooltip_title");
        const iconClassName =
            kind !== InfoTooltipKind.Warning ? "mx_InfoTooltip_icon_info" : "mx_InfoTooltip_icon_warning";

        // Tooltip are forced on the right for a more natural feel to them on info icons
        return (
            <Tooltip label={tooltip || title} placement="right">
                <div className={classNames("mx_InfoTooltip", className)} tabIndex={this.props.tabIndex ?? 0}>
                    <span className={classNames("mx_InfoTooltip_icon", iconClassName)} aria-label={title} />
                    {children}
                </div>
            </Tooltip>
        );
    }
}
