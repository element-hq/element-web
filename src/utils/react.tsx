/*
Copyright 2024 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only
Please see LICENSE files in the repository root for full details.
*/

import { ReactNode } from "react";
import { createRoot, Root } from "react-dom/client";

/**
 * Utility class to render & unmount additional React roots,
 * e.g. for pills, tooltips and other components rendered atop user-generated events.
 */
export class ReactRootManager {
    private roots: Root[] = [];
    private rootElements: Element[] = [];

    public get elements(): Element[] {
        return this.rootElements;
    }

    public render(children: ReactNode, element: Element): void {
        const root = createRoot(element);
        this.roots.push(root);
        this.rootElements.push(element);
        root.render(children);
    }

    public unmount(): void {
        while (this.roots.length) {
            const root = this.roots.pop()!;
            this.rootElements.pop();
            root.unmount();
        }
    }
}
