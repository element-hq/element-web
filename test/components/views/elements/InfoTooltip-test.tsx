/*
Copyright 2023 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";
import userEvent from "@testing-library/user-event";
import { render, waitFor } from "@testing-library/react";

import InfoTooltip from "../../../../src/components/views/elements/InfoTooltip";

describe("InfoTooltip", () => {
    it("should show tooltip on hover", async () => {
        const { getByText, asFragment } = render(<InfoTooltip tooltip="Tooltip text">Trigger text</InfoTooltip>);

        const trigger = getByText("Trigger text");
        expect(trigger).toBeVisible();
        await userEvent.hover(trigger!);

        // wait for the tooltip to open
        const tooltip = await waitFor(() => {
            const tooltip = document.getElementById(trigger.getAttribute("aria-describedby")!);
            expect(tooltip).toBeVisible();
            return tooltip;
        });
        expect(tooltip).toHaveTextContent("Tooltip text");

        expect(asFragment()).toMatchSnapshot();
    });
});
