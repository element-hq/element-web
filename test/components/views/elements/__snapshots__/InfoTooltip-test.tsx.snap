// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfoTooltip should show tooltip on hover 1`] = `
<DocumentFragment>
  <div
    aria-describedby="floating-ui-2"
    class="mx_InfoTooltip"
    tabindex="0"
  >
    <span
      aria-label="Information"
      class="mx_InfoTooltip_icon mx_InfoTooltip_icon_info"
    />
    Trigger text
  </div>
</DocumentFragment>
`;
